hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+
+                # Check to see if anything has come through the channel
+                self.read_rtm_channel()
+
+                # Run any scheduled tasks
+                schedule.run_pending()
+
+                # Sleep for a bit until the next time we read the channel
+                time.sleep(POLLING_INTERVAL)
+
+        else:
+            raise SlackException(
+                "could not connect to API: invalid Slack API key or Bot ID?"
+            )
+
+        # If we've made it here, we've been shutdown
+        # TODO: convert to logging functionality
+        print("bot has been gracefully shutdown")
+
+    def stop(self):
+        """
+        Set the shutdown semaphore
+        """
+        self.shutdown = True
+
+    def read_rtm_channel(self):
+        """
+        Reads the RTM channel, parses output, filters messages based on type
+        and if they are directed at the bot. If so, passes those messages on
+        to the message handler command.
+        """
+
+        # Check for messages on the wire
+        # TODO: as soon as we connect, the last message comes through.
+        messages = self.slack.rtm_read()
+        if messages and len(messages) > 0:
+            # If there are messages, begin filtering and parsing.
+            for msg in messages:
+                # Filter only messages from other users
+                if msg["type"] == "message":
+                    # Check if an @botid appears in the message
+                    if self.atbotid.search(msg.get("text", "")):
+                        # Handle the atbotid message
+                        self.handle_message(msg)
+
+    def handle_message(self, msg):
+        """
+        Handles messages that are directed at the bot by parsing commands.
+        """
+        # Log the message received
+        # TODO: change to standardized logging functionality
+        print("[{}] {}".format(msg["ts"], msg["text"]))
+
+        # Track if an unknown command has occurred
+        unknown = True
+
+        # Determine if the message matches any commands
+        for name, command in self.commands.items():
+            match = command.search(msg["text"])
+            if match is not None:
+                unknown = False # We've found at least one command match.
+
+                # Get the method that handles this command
+                handler = getattr(self, "handle_{}_command".format(name), None)
+                if handler is None:
+                    raise NotImplementedError(
+                        "{} command not yet handled".format(name)
+                    )
+
+                try:
+                    # Handle the command
+                    handler(msg)
+                except HanishException as e:
+                    # Respond with hanish exceptions
+                    response = "Sorry, there was a problem with your request: {}".format(e)
+                    self.post(msg['channel'], response)
+                except Exception as e:
```

Comment:
Yes, at this level, user-level exceptions are the only thing that are handled, and that was most convenient for the initial prototype because I didn't know the kind of exceptions I would get from the various APIs. Before production I would recommend that the body of the `while` loop in the `bot.run` method is wrapped by `try/except`, however initialization exceptions (trying to connect to dark sky, finding required configurations, connecting to the Slack API) all remain uncaught as a signal that something is wrong at startup. 

Even with exception handling at runtime, I would probably still kill the process (though more gracefully, with logging for example) and manage the service with upstart or supervisord. The external processes could then attempt restarts, with delays in between to cover outages, but limit the number of restarts before alerting the admins. 

So briefly, here are the merits/disadvantages just as a start to the discussion 

Merits:

- Allows OS or external programs to assist in process management
- The ability to handle control flow at any level of the call stack 
- Ability to provide different exception handling logic in tiers (e.g. initialization vs. run vs. message handling) 
- Functional contracts - "this is the input I expect, this is the output you can expect", exceptions are a "breach of contract" 
- Exceptions can be passed as signals in multi-threaded or multi-process programs (otherwise shared state or synchronization barriers are required, which can be difficult) 
- Exceptions are data types that can be passed to other processes (e.g. for logging) 

Disadvantages:

- Makes things less clear about where program control is handled (have to hold the call stack in your head while programming)
- Catching all exceptions can hide a huge amount of problems 
- Exceptions may not be defined as something that can be returned from a specific function (e.g. can come from lower in the call stack, this type of definition is actually something I hated about Java originally, but then grew to appreciate)